src/TreeSupport.cpp

```diff
@@ -650,32 +610,119 @@ void TreeSupport::insertDroppedNode(std::unordered_set<Node*>& nodes_layer, Node
-            previous_layer = previous_layer.unionPolygons(model_collision[radius_sample][layer_nr]);
-            model_avoidance[radius_sample].push_back(previous_layer);
-        }
-#pragma omp atomic
-        completed++;
-#pragma omp critical (progress)
-        {
-            Progress::messageProgress(Progress::Stage::SUPPORT, ((model_collision.size() / 2) + (completed / 2)) * PROGRESS_WEIGHT_COLLISION, model_avoidance.size() * PROGRESS_WEIGHT_COLLISION + storage.support.supportLayers.size() * PROGRESS_WEIGHT_DROPDOWN + storage.support.supportLayers.size() * PROGRESS_WEIGHT_AREAS);
-        }
+        return it->second;
+    }
+    else
+    {
+        return calculateCollision(key);
     }
 }
 
+const Polygons& ModelVolumes::avoidance(coord_t radius, int layer) const
+{
+    radius = roundRadius(radius);
+    RadiusLayerPair key{radius, layer};
+    const auto it = avoidance_cache_.find(key);
+    if (it != avoidance_cache_.end())
+    {
+        return it->second;
+    }
+    else
+    {
+        return calculateAvoidance(key);
+    }
+}
+
+const Polygons& ModelVolumes::internal_model(coord_t radius, int layer) const
+{
+    radius = roundRadius(radius);
+    RadiusLayerPair key{radius, layer};
+    const auto it = internal_model_cache_.find(key);
+    if (it != internal_model_cache_.end())
+    {
+        return it->second;
+    }
+    else
+    {
+        return calculateInternalModel(key);
+    }
+}
+coord_t ModelVolumes::roundRadius(coord_t radius) const
+{
+    const auto remainder = radius % radius_sample_resolution_;
+    const auto delta = remainder != 0 ? radius_sample_resolution_- remainder : 0;
+    return radius + delta;
+}
+
+const Polygons& ModelVolumes::calculateCollision(const RadiusLayerPair& key) const
+{
+    const auto& radius = key.first;
+    const auto& layer = key.second;
+
+    auto collision_areas = machine_border_;
+    if (layer < static_cast<int>(layer_outlines_.size())) {
+        collision_areas = collision_areas.unionPolygons(layer_outlines_[layer]);
+    }
+    collision_areas = collision_areas.offset(xy_distance_ + radius, ClipperLib::JoinType::jtRound);
+    const auto ret = collision_cache_.insert({key, std::move(collision_areas)});
+    assert(ret.second);
+    return ret.first->second;
+}
+
+const Polygons& ModelVolumes::calculateAvoidance(const RadiusLayerPair& key) const
+{
+    const auto& radius = key.first;
+    const auto& layer = key.second;
+
+    if (layer == 0)
+    {
+        avoidance_cache_[key] = collision(radius, 0);
+        return avoidance_cache_[key];
+    }
+
+    // Avoidance for a given layer depends on all layers beneath it so could have very deep recursion depths if
+    // called at high layer heights
+    constexpr auto max_recursion_depth = 100;
```

Comment:
I find it hard to understand this piece of code.
Please add more comments in the code to explain what you are doing here.

It took me 15 minutes and a rubber duck to understand! (Am I dumb?)

I think it's not too important to prevent high recursion depths. You're only putting two ints and a pointer on the stack for every call.

Perhaps it's best to rewrite this code so that it just uses a for-loop instead of recursion.

Reply:
The stack frames are small but i've been bitten by this before so wanted to be a bit defensive.
I'll add some comments to explain whats going on.

Reply:
Why not use a for-loop isntead of recursion?

Reply:
It complicates the logic a bit because we need to either loop backwards from `layer_idx` until we find the highest layer in the cache or loop upwards from zero and then run the loop forwards. Something like this (theres probably an off-by-one error in here):

```suggestion
    auto prev_layer = layer_idx - 1;
    for (; prev_layer > 0; --prev_layer)
    {
        if (avoidance_cache_.find({radius, prev_layer}) != avoidance_cache_.end())
        {
            break;
        }
    }
    for (; prev_layer < layer_idx + 1; ++prev_layer)
    {
        auto avoidance_areas = avoidance_cache_[{radius, prev_layer}].offset(-max_move_).smooth(5);
        avoidance_areas = avoidance_areas.unionPolygons(getCollision(radius, prev_layer + 1));
        const auto ret = avoidance_cache_.insert({key, std::move(avoidance_areas)});
    }

    return avoidance_cache_[key];
```

This isn't as clean to my mind but i guess the recursive solution isn't particularly clean when the defensive part is included. I'll leave it up to you, if the for loop is the preference i can do that instead.